api/jwt.go

```diff
@@ -3,77 +3,227 @@ package api
+	}
+	return oathkeeperPubKeys, nil
+}
+
+// tokenFromRequest extracts the JWT token from the request and returns it.
+// Returns an empty string if there is no token.
+func tokenFromRequest(r *http.Request) string {
+	// Check the headers for a token.
 	h := r.Header.Get("Authorization")
-	strArr := strings.Split(h, " ")
-	if len(strArr) == 2 {
-		return strArr[1]
+	if !strings.HasPrefix(h, "Bearer ") {
+		return ""
+	}
+	return strings.TrimPrefix(h, "Bearer ")
+}
+
+// tokenFromContext is a helper function that extracts the JWT token from the
+// context and returns the contained user sub, claims and the token itself.
+// The sub is the user id used in Kratos.
+//
+// Example claims structure:
+//
+// map[
+//    exp:1.607594172e+09
+//    iat:1.607593272e+09
+//    iss:https://siasky.xyz/
+//    jti:1e5872ae-71d8-49ec-a550-4fc6163cbbf2
+//    nbf:1.607593272e+09
+//    sub:695725d4-a345-4e68-919a-7395cb68484c
+//    session:map[
+//        active:true
+//        authenticated_at:2020-12-09T16:09:35.004003Z
+//        issued_at:2020-12-09T16:09:35.004042Z
+//        expires_at:2020-12-10T16:09:35.004003Z
+//        id:9911ad26-e47f-4ec4-86a1-fbbc7fd5073e
+//        identity:map[
+//            id:695725d4-a345-4e68-919a-7395cb68484c
+//            recovery_addresses:[
+//                map[
+//                    id:e2d847e1-1885-4edf-bccb-64b527b30096
+//                    value:ivaylo@nebulous.tech
+//                    via:email
+//                ]
+//            ]
+//            schema_id:default
+//            schema_url:https://siasky.xyz/secure/.ory/kratos/public/schemas/default
+//            traits:map[
+//                email:ivaylo@nebulous.tech
+//                name:map[
+//                    first:Ivaylo
+//                    last:Novakov
+//                ]
+//            ]
+//            verifiable_addresses:[
+//                map[
+//                    id:953b0c1a-def9-4fa2-af23-fb36c00768d2
+//                    status:pending
+//                    value:ivaylo@nebulous.tech
+//                    verified:true
+//                    verified_at:2020-12-09T16:09:35.004042Z
+//                    via:email
+//                ]
+//            ]
+//        ]
+//    ]
+// ]
+func tokenFromContext(req *http.Request) (sub string, claims jwt.MapClaims, token *jwt.Token, err error) {
+	t, ok := req.Context().Value(ctxValue("token")).(*jwt.Token)
+	if !ok {
+		err = errors.New("failed to get token")
+		return
 	}
-	return ""
+	if reflect.ValueOf(t.Claims).Kind() != reflect.ValueOf(jwt.MapClaims{}).Kind() {
+		err = errors.New("the token does not contain the claims we expect")
+		return
+	}
+	claims = t.Claims.(jwt.MapClaims)
+	if reflect.ValueOf(claims["sub"]).Kind() != reflect.String {
+		err = errors.New("the token does not contain the sub we expect")
+	}
+	defer func() {
+		if e := recover(); e != nil {
+			sub = ""
+			claims = nil
+			token = nil
+			err = errors.New("jwt claims don't contain a valid sub")
```

Comment:
I think we'd want to include `e` here in the error message right?

Reply:
Wait why can't we just check if the field exists? e.g.
```go
subEntry, exists := claims["sub"]
if !exists {
  return errors.New("claim doesn't contain a sub")
}
sub = subEntry.(string)
```
?

We should unnecessary `recover` defers in production whenever possible.

Reply:
Oh, good idea.

The only thing is whether we can be sure that if it's there it would be a string and not, say, an int? If the type is defined as `map[string]interface{}` and the value is `{"foo":1}` then it might be an int, right?

It seems safer to just check it with `reflect`:
```go
	if reflect.ValueOf(claims["sub"]).Kind() != reflect.String {
		claims = nil
		err = errors.New("jwt claims don't contain a valid sub")
		return
	}
	sub = claims["sub"].(string)
```


OTOH, this is a validated token coming from Kratos and we know the `sub` will be a string. I'll just use your approach, because it's simpler.